pkg/restore/restore.go

```diff
@@ -979,122 +984,167 @@ func (ctx *restoreContext) restoreItem(obj *unstructured.Unstructured, groupReso
 		return warnings, errs
 	}
 
-	if groupResource == kuberesource.PersistentVolumes {
-		switch {
-		case hasSnapshot(name, ctx.volumeSnapshots):
-			oldName := obj.GetName()
-			shouldRenamePV, err := shouldRenamePV(ctx, obj, resourceClient)
+	var itemSnapshot *volume.ItemSnapshot
+	if features.IsEnabled(velerov1api.UploadProgressFeatureFlag) { // Use the ResourceIdentifier for the original item to match for snapshots
+		itemRID := velero.ResourceIdentifier{
+			GroupResource: groupResource,
+			Namespace:     obj.GetNamespace(),
+			Name:          name,
+		}
+		itemSnapshot, err = getItemSnapshot(itemRID, ctx.itemSnapshots)
+		if err != nil {
+			errs.Add(namespace, err)
+			return warnings, errs
+		}
+	}
+	if itemSnapshot != nil {
+		itemSnapshotter := clientmgmt.ItemSnapshotterForSnapshotFromResolved(itemSnapshot, ctx.itemSnapshotterActions)
+		if itemSnapshotter != nil {
+			cii := isv1.CreateItemInput{
+				SnapshottedItem:  obj,
+				SnapshotID:       itemSnapshot.Status.ProviderSnapshotID,
+				ItemFromBackup:   obj,
+				SnapshotMetadata: itemSnapshot.Status.Metadata,
+				Params:           nil, // TBD
+				Restore:          ctx.restore,
+			}
+			cio, err := itemSnapshotter.CreateItemFromSnapshot(go_context.TODO(), &cii)
 			if err != nil {
 				errs.Add(namespace, err)
 				return warnings, errs
 			}
-
-			// Check to see if the claimRef.namespace field needs to be remapped,
-			// and do so if necessary.
-			_, err = remapClaimRefNS(ctx, obj)
-			if err != nil {
-				errs.Add(namespace, err)
+			unstructuredObj, ok := cio.UpdatedItem.(*unstructured.Unstructured)
+			if !ok {
+				errs.Add(namespace, fmt.Errorf("%s: unexpected type %T", resourceID, cio.UpdatedItem))
 				return warnings, errs
 			}
 
-			var shouldRestoreSnapshot bool
-			if !shouldRenamePV {
-				// Check if the PV exists in the cluster before attempting to create
-				// a volume from the snapshot, in order to avoid orphaned volumes (GH #609)
-				shouldRestoreSnapshot, err = ctx.shouldRestore(name, resourceClient)
+			obj = unstructuredObj
+			w, e := handleAdditionalItems(ctx, namespace, cio.AdditionalItems)
+			warnings.Merge(&w)
+			errs.Merge(&e)
+		} else {
+			errs.Add(namespace, fmt.Errorf("could not find ItemSnapshotter %s needed for snapshot of item %s, snapshot ID %s",
+				itemSnapshot.Spec.ItemSnapshotter, itemSnapshot.Spec.ResourceIdentifier, itemSnapshot.Status.ProviderSnapshotID))
+		}
+	} else {
```

Comment:
I feel this flow may need to be revisited.

If the current obj is a PV and it has a itemSnapshot, after the PV is restored by the plugin, the additional changes to the PV resource like renaming, reference remapping will not be executed.

I think the changes should be made in the velero process, not the plugin.

Reply:
If we do that in Velero then we have essentially hardcoded the logic for PVs into Velero.  That's where we're at now and it's what we want to get away from.  Renaming and reference remapping will need to be handled for other types as well and it is type specific, not generic logic.

Reply:
Sorry but I still have some concerns about this:

If multiple there are itemsnapshotters taking snapshots for the same type of resources like PV, there will be duplication of code in the different plugins.

Moreover, I used to think the itemsnapshotter plugin should be focusing on snapshotting or restoring resources, but the condition for renaming PVs is based on the context of the restore, this makes the itemsnapshotter harder to maintain b/c it will be more tightly coupled to the overall restore workflow.

Reply:
So the thought I had was to have one PVC snapshotter, which would be a built-in plugin to Velero, the way that some of the existing BackupItemActions are set up.  This could use CSI snapshots where CSI snapshotting was available.  For system specific things, that would go through Astrolabe or the PVC ItemSnapshotter could call the existing VolumeSnapshot plugins for system specific stuff.  Existing VolumeSnapshot plugins do not have data path APIs, however, so eventually that would need to be added, which was one of the reasons to move to Astrolabe which is type agnostic, but I'll leave that to you.  All PVCs would go through the built-in PVC ItemSnapshotter.  Existing CSI BackupItemAction plugin and vSphere BackupItemAction duplicate code and are different.  Unifying the CSI portions into a single plugin will make it better.

If you want to do it differently, be my guest, but I don't see how it would all work together.  Splitting between ItemSnapshotter and BackupItemAction will be hard to co-ordinate.  Coding PV specific info into Velero means that additional types such as VirtualMachine which also require modification of the Kubernetes resource will need hard-coding into Velero.